File: sudoku.jl
# create the colormap for a traditional sudoku, where a 9*9 board is split into 9 3*3 areas of the same colour
function createNormalSudokuMap()
	out = zeros(Int8,9,9)
	inc = [1 1 1 2 2 2 3 3 3]

	for i in [1,4,7]
		big = [inc;inc;inc]
		inc = inc .+ 3
		out[i:i+2,:] = big
	end

	return out
end

#=
	sudoku -> the sudoku grid, can have some numbers already filled
	sudoku_map -> the colormap for the sudoku, a color cannot have two numbers with the same value
	number -> the number we are trying to insert
	cartesian -> (i,j) touple, where i is the row index and j is the column index
=#
function checkForAvailability(sudoku, sudoku_map, number, cartesian::CartesianIndex{2})
	function checkRowAva(number, row)
		if number in row
			return false
		end
		return true
	end

	row_available = checkRowAva(number,sudoku[cartesian[1],:])
	column_available = checkRowAva(number,sudoku[:,cartesian[2]])

	colour = sudoku_map[cartesian]
	result = map(x -> x == number, sudoku)
	colors = map(x -> x == colour, sudoku_map)

	ultraRes = result .* colors

	if (row_available && column_available && !(true in ultraRes))
		return true
	end
	return false
end

# creates a new sudoku based on the sudoku_map, with "filled" fields filled, (filled is a number, that many fields remain filled)
function createSudoku(sudoku_map,filled = 17)
	sudoku = zeros(Int8,size(sudoku_map))
	sudoku_heuristic = createHeuristicRandom(sudoku_map)

	sudoku = solveSudokuHeuristic(sudoku,sudoku_map,sudoku_heuristic)[3]
	to_leave = zeros(Int8,size(sudoku_map))

	while filled >= 0
		i = rand(1:maximum(size(sudoku)))
		j = rand(1:maximum(size(sudoku)))
		
		if to_leave[i,j] != 1
			to_leave[i,j] = 1
			filled = filled - 1
		end
	end

	solution = copy(sudoku)
	sudoku = sudoku .* to_leave

	return (sudoku,solution)
end

function createHeuristicRandom(sudoku_map,filled = 16)
	sudoku_heuristic = ones(Int8,size(sudoku_map))
	while filled >= 0
		i = rand(1:maximum(size(sudoku_map)))
		j = rand(1:maximum(size(sudoku_map)))
		
		if sudoku_heuristic[i,j] != 2
			sudoku_heuristic[i,j] = 2
			filled = filled - 1
		end
	end
	return sudoku_heuristic
end

#=
	sudoku -> the sudoku grid, can have some numbers already filled
	sudoku_map -> the colormap for the sudoku, a color cannot have two numbers with the same value
	sudoku_heuristic -> optional, which fields the sudoku solver prioritizes, if not provided all fields have an equal pripority, the solver will solve up to down, left to right
	returns (solution_found, number of steps, sudoku)
		solution_found -> boolean, weather a solution was found or not
		number of steps -> number of different numbers tried to arrive to the returned sudoku
		sudoku -> the sudoku the program had stored at the end
=#
function solveSudokuHeuristic(sudoku,sudoku_map,sudoku_heuristic = nothing)
	function is_zero(x)
		if x == 0
			return 1
		end
		return 0
	end

	if sudoku_heuristic == nothing
		sudoku_heuristic = map(is_zero,sudoku)
	else
		values = map(x -> x == 0,sudoku)
		sudoku_heuristic = sudoku_heuristic .* values
	end

	start = findmax(sudoku_heuristic)[2]
	return solveSudokuRek(sudoku,sudoku_map,sudoku_heuristic,start,Int64(0))
end

#=
	helper function for function solveSudokuHeuristic
	curr -> current field being filled, type CartesianIndex(i, j)
=#
function solveSudokuRek(sudoku,sudoku_map,sudoku_heuristic,curr,num_steps)
	if num_steps < 0
		@warn "overflow for number of steps, result will be inaccurate", num_steps
	elseif num_steps > 1000000
		error("exceded 1000000 steps", sudoku,sudoku_heuristic)
	end
	
	heuristic = sudoku_heuristic[curr]
	for number in 1:maximum(size(sudoku))
		num_steps += 1
		if checkForAvailability(sudoku,sudoku_map,number,curr)
			sudoku[curr] = number
			sudoku_heuristic[curr] = 0
			next = findmax(sudoku_heuristic)
			
			# solved sudoku
			if next[1] == 0
				return (true,num_steps,sudoku)
			end
			
			solution = solveSudokuRek(sudoku,sudoku_map,sudoku_heuristic,next[2],num_steps)

			# if we found a solution
			if solution[1]
				return solution
			# if we didn't find a solution, change the number of steps to that of the subtree
			else
				num_steps = solution[2]
			end
			# cleanup if a solution was not found
			sudoku_heuristic[curr] = heuristic
			sudoku[curr] = 0
		end
	end
	return (false,num_steps,sudoku)
end

sudoku_map = createNormalSudokuMap()
sudoku = createSudoku(sudoku_map)
println(sudoku)

File: sudoku.py
import numpy as np
import random as r

def create_normal_sudoku_map():
    out = np.zeros((9,9),np.int8)
    inc = np.array([[1,1,1,2,2,2,3,3,3]])
    
    for i in [0,3,6]:
        big = np.concatenate((inc,inc,inc),axis=0)
        out[i:i+3,:] = big
        inc = np.add(inc,3)
    return out

def check_row_ava(number,row):
    if number in row:
        return False
    return True

def check_for_availability(sudoku,sudoku_map,number,coords):
    row_available = check_row_ava(number,sudoku[coords[0],:])
    column_available = check_row_ava(number,sudoku[:,coords[1]])
    
    colour = sudoku_map[coords]
    
    result = (lambda x: x == number)(sudoku)
    colours = (lambda x: x == colour)(sudoku_map)
    
    same_colour_same_number = np.multiply(result,colours)
    
    if(row_available and column_available and True not in same_colour_same_number):
        return True
    return False

def create_heuristic_random(sudoku_map,filled = 17):
    sudoku_heuristic = np.ones(sudoku_map.shape,np.int8)
    length = max(sudoku_map.shape)
    
    while filled >= 0:
        i = r.randint(0,length - 1)
        j = r.randint(0,length - 1)
        
        if sudoku_heuristic[i,j] != 2:
            sudoku_heuristic[i,j] = 2
            filled -= 1
    return sudoku_heuristic

def solve_sudoku_heuristic(sudoku,sudoku_map,sudoku_heuristic = None):
    if sudoku_heuristic is None:
        sudoku_heuristic = np.ones(sudoku.shape,np.int8)
    
    # write zeros where there are numbers already filled in
    solutions = (lambda x: x== 0)(sudoku)
    sudoku_heuristic = np.multiply(solutions,sudoku_heuristic)
    
    start = np.unravel_index(np.argmax(sudoku_heuristic),sudoku_heuristic.shape)
    return solve_sudoku_rek(sudoku,sudoku_map,sudoku_heuristic,start,0)

def solve_sudoku_rek(sudoku,sudoku_map,sudoku_heuristic,curr,num_steps):
    if num_steps < 0:
        print("WARNING: number of steps overflowed, result will not be accurate")
    elif num_steps > 1000000:
        #print (sudoku_heuristic)
        #print()
        #print(sudoku)
        raise AssertionError("number of steps exceeded 1000000")
    
    heuristic = sudoku_heuristic[curr]
    
    for number in range(1,max(sudoku.shape)+1):
        num_steps += 1
        if check_for_availability(sudoku,sudoku_map,number,curr):
            sudoku[curr] = number
            sudoku_heuristic[curr] = 0
            nex = np.unravel_index(np.argmax(sudoku_heuristic),sudoku_heuristic.shape)
            
            if np.amax(sudoku_heuristic) == 0:
                return (True,num_steps,sudoku)
            
            solution = solve_sudoku_rek(sudoku,sudoku_map,sudoku_heuristic,nex,num_steps)
            
            if solution[0]:
                return solution
            else:
                num_steps = solution[1]
            
            sudoku_heuristic[curr] = heuristic
            sudoku[curr] = 0
    return (False,num_steps,sudoku)

def create_sudoku(sudoku_map,filled = 17):
    sudoku = np.zeros(sudoku_map.shape,np.int8)
    sudoku_heuristic = create_heuristic_random(sudoku_map)
    
    solution = solve_sudoku_heuristic(sudoku,sudoku_map,sudoku_heuristic)[2]
    length = max(sudoku_map.shape)
    
    to_leave = np.zeros(sudoku_map.shape,np.int8)
    
    while filled >= 0:
        i = r.randint(0,length - 1)
        j = r.randint(0,length - 1)
        
        if to_leave[i,j] != 1:
            to_leave[i,j] = 1
            filled -= 1
    sudoku = np.multiply(solution,to_leave)
    return (sudoku,solution)
    
    
if __name__ == "__main__":
    sudoku = np.zeros((9,9),np.int8)
    sudoku[0,0] = 5
    sudoku_map = create_normal_sudoku_map()
    